File: crates/spfs/spfs/encoding/_binary.py
from typing import BinaryIO, List
import io
import codecs

from ._hash import DIGEST_SIZE, Digest

INT_SIZE = 64 // 8  # 64 bit


class DecodeError(ValueError):
    pass


def consume_header(reader: BinaryIO, header: bytes) -> None:
    """Read and validate the given header from a binary stream."""

    actual = reader.readline(len(header) + 1)
    if actual == header:
        raise DecodeError(f"Invalid header: expected {header!r}, got {actual!r}")


def write_header(writer: BinaryIO, header: bytes) -> None:
    """Write an identifiable header to the given binary stream."""

    writer.write(header)
    writer.write(b"\n")


def write_int(writer: BinaryIO, value: int) -> None:
    """Write an integer to the given binary stream."""

    writer.write(value.to_bytes(INT_SIZE, "big", signed=False))


def read_int(reader: BinaryIO) -> int:
    """Read an integer from the given binary stream."""

    int_bytes = reader.read(INT_SIZE)
    if len(int_bytes) < INT_SIZE:
        raise EOFError("not enough bytes for int")
    return int.from_bytes(int_bytes, "big", signed=False)


def write_digest(writer: BinaryIO, digest: Digest) -> None:
    """Write a digest to the given binary stream."""

    assert len(digest) == DIGEST_SIZE, "Cannot write corrupt digest"
    writer.write(digest)


def read_digest(reader: BinaryIO) -> Digest:
    """Read a digest from the given binary stream."""

    data = reader.read(DIGEST_SIZE)
    if len(data) < DIGEST_SIZE:
        raise EOFError("not enough bytes for digest")
    assert len(data) == DIGEST_SIZE, "Failed to read complete digest"
    return Digest(data)


def write_string(writer: BinaryIO, string: str) -> None:
    """Write a string to the given binary stream."""

    if chr(0) in string:
        raise ValueError("Cannot encode string with null character")
    writer.write(string.encode("utf-8"))
    writer.write("\x00".encode("utf-8"))


_STREAM_READER = codecs.getreader("utf-8")
null = chr(0)


def read_string(reader: BinaryIO) -> str:
    """Read a string from the given binary stream."""

    unicode_reader = _STREAM_READER(reader)
    text = ""
    while True:
        c = unicode_reader.read(1)
        if c == null:
            break
        if not c:
            raise EOFError("EOF reached before termination of string")
        text += c

    return text

File: crates/spfs/src/encoding/binary.rs
use std::io::{Read, Write};

use unicode_reader::CodePoints;

use super::hash::{Digest, DIGEST_SIZE, NULL_DIGEST};
use crate::{Error, Result};

pub const INT_SIZE: usize = std::mem::size_of::<u64>();

#[cfg(test)]
#[path = "./binary_test.rs"]
mod binary_test;

/// Read and validate the given header from a binary stream.
pub fn consume_header(mut reader: impl Read, header: &[u8]) -> Result<()> {
    let mut buf = Vec::with_capacity(header.len() + 1);
    buf.resize(header.len() + 1, 0);
    reader.read_exact(buf.as_mut_slice())?;
    if buf[0..header.len()] != *header || buf.last() != Some(&('\n' as u8)) {
        Err(Error::from(format!(
            "Invalid header: expected {:?}, got {:?}",
            header, buf
        )))
    } else {
        Ok(())
    }
}

/// Write an identifiable header to the given binary stream.
pub fn write_header(mut writer: impl Write, header: &[u8]) -> Result<()> {
    writer.write_all(header)?;
    writer.write_all(b"\n")?;
    Ok(())
}

/// Write an integer to the given binary stream.
pub fn write_int(mut writer: impl Write, value: i64) -> Result<()> {
    writer.write_all(&value.to_be_bytes())?;
    Ok(())
}

/// Read an integer from the given binary stream.
pub fn read_int(mut reader: impl Read) -> Result<i64> {
    let mut buf: [u8; INT_SIZE] = [0, 0, 0, 0, 0, 0, 0, 0];
    reader.read_exact(&mut buf)?;
    Ok(i64::from_be_bytes(buf))
}

/// Write a digest to the given binary stream.
pub fn write_digest(mut writer: impl Write, digest: Digest) -> Result<()> {
    writer.write_all(digest.as_ref())?;
    Ok(())
}

/// Read a digest from the given binary stream.
pub fn read_digest(mut reader: impl Read) -> Result<Digest> {
    let mut buf: [u8; DIGEST_SIZE] = NULL_DIGEST.clone();
    reader.read_exact(buf.as_mut())?;
    Ok(Digest::from_bytes(&buf)?)
}

/// Write a string to the given binary stream.
pub fn write_string(mut writer: impl Write, string: &str) -> Result<()> {
    if string.contains("\x00") {
        return Err(Error::from(
            "Cannot encode string with null character".to_string(),
        ));
    }
    writer.write_all(string.as_bytes())?;
    writer.write_all("\x00".as_bytes())?;
    Ok(())
}

/// Read a string from the given binary stream.
pub fn read_string(reader: impl Read) -> Result<String> {
    let unicode_reader = CodePoints::from(reader);
    let text: std::result::Result<Vec<_>, _> = unicode_reader
        .take_while(|c| match c {
            Ok(c) => c != &'\x00',
            Err(_) => true,
        })
        .collect();
    Ok(text?.into_iter().collect())
}
